File: 21558 - Nithin/week 8/81-search-in-rotated-sorted-array-ii/81-search-in-rotated-sorted-array-ii.cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int n = nums.size();
        int left = 0, right = n-1;
        int pos = -1, mid;
        
        while(left <= right){
            mid = (left+right)/2;
    
            if(nums[mid] == target)
                return true;
            
            else if(nums[left] == nums[mid] && nums[right] == nums[mid]){
                left++;
                right--;
            }
            
            else if (nums[left] <= nums[mid]){
                if(target < nums[mid] && target >= nums[left])
                    right = mid-1;
                else
                    left = mid+1;
            }
        
            else{
                if(target > nums[mid] && target <= nums[right])
                    left = mid+1;
                else
                    right = mid-1;
            }
        }
        return false;
    }
};

File: 21551-naveen-siva-kumar/week8/81_Search in Rotated Sorted Array II.py
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int left =0;
        int right = nums.size()-1;
        int mid =0;
        
        while(left<=right){
            
            mid = (left+right)/2;
            
            if(nums[mid] == target) return true;
            
            else if(nums[mid]>nums[right]){
                if(nums[left]<=target && nums[mid]>target){ right = mid-1;}
                else {left =mid+1;}
                //cout<<"right "<<right<<"left "<<left<<endl;
            }
            
            else if(nums[mid]<nums[right]){
                if(nums[right]>=target && nums[mid]<target){left = mid+1;}
                else {
                    right = mid-1;
                }
                //cout<<"right "<<right<<"left "<<left<<endl;
            }
            else{right--;}
                
        
        }
        return false;
    }
};
